querier.go

```diff
@@ -1060,30 +1060,50 @@ func (it *verticalMergeSeriesIterator) Err() error {
 type chunkSeriesIterator struct {
 	chunks []chunks.Meta
 
-	i   int
-	cur chunkenc.Iterator
+	i          int
+	cur        chunkenc.Iterator
+	bufXorIter *chunkenc.XORIterator
+	bufDelIter *deletedIterator
 
 	maxt, mint int64
 
 	intervals Intervals
 }
 
 func newChunkSeriesIterator(cs []chunks.Meta, dranges Intervals, mint, maxt int64) *chunkSeriesIterator {
-	it := cs[0].Chunk.Iterator()
-
-	if len(dranges) > 0 {
-		it = &deletedIterator{it: it, intervals: dranges}
-	}
-	return &chunkSeriesIterator{
+	csi := &chunkSeriesIterator{
 		chunks: cs,
 		i:      0,
-		cur:    it,
 
 		mint: mint,
 		maxt: maxt,
 
 		intervals: dranges,
 	}
+	csi.resetCurIterator()
+
+	return csi
+}
+
+func (it *chunkSeriesIterator) resetCurIterator() {
+	if it.bufXorIter != nil {
+		it.bufXorIter.Reset(it.chunks[it.i].Chunk.Bytes())
```

Comment:
Aren't you presuming the type here?

Reply:
If 1 chunk was found to be of a certain type, would it be wrong to assume the further chunks in the slice would be different?

Reply:
Yes, that's not safe in general. We're looking at new types of encoding right now for example.

Reply:
Yes, that's right. Then this whole `Reset()` method might not be required as using it for the current use case is not clean (even the use case that I had in cortex with this was similar to what I was doing here).